Security controls > Tokenization docs (#64)
cossacklabs/product-docs: acra/security-controls/tokenisation/_index.md

title: Tokenisation
bookCollapseSection: true
---

# Tokenization

## Description

Tokenization is a way to transform some sensitive data into pseudo-random representation of same data type.
A number will remain a number, string will be string, email will be email.
But the tokenized version will look like a random value.

> Better see than read:

Reviewer: ```suggestion
It is better to see than to read:
```

* `789` -> `156749362`
* `testing testing` -> `lwcIvM3zaJN9cbN`

> * `kate@example.com` -> `sVAOfOwF@mSuGPog.et`

Reviewer: It seems that email is not data type. I guess that tokenization targets not only data type but also format (email will be email). Should we elaborate on this in phrase above?

The tokenization process has few interesting properties:

* Tokenized value is simply random, it is neither encrypted nor hashed version of the input

> * Tokenization of the same value more than once will return same result

Reviewer 2: may return the same result (in consistent mode) or random (not consistent) even for the same value according to configuration

> * Tokenization of the same value more than once will return same result

Reviewer 3: ```suggestion
* Tokenization supports two modes: tokenization of the same value will return the same result every time (called `"consistent tokenization"`). Tokenization of the same value will return different results every time (called `"inconsistent tokenization"`.
```

* Tokenization can be reversed, but only for valid (previously returned) values

## API

> Tokenization is done by Acra Translator.

Reviewer 2: AcraServer tokenize values too according to `encryptor_config_file`

Both protocols (gRPC and HTTP) provide the same functionality, but from different angles.
You decide which one fits best for your application.

### gRPC

In order to use the translator gRPC API you have to take
[api.proto](https://github.com/cossacklabs/acra-Q12021/blob/master/cmd/acra-translator/grpc_api/api.proto)

> and use either `service Tokenizator` or `service BulkProcessing` (with one one more `TokenizeRequest` inside).

Reviewer: ```suggestion
and use either `service Tokenizator` or `service BulkProcessing` (with one more `TokenizeRequest` inside).
```

#### Golang example

> TODO add golang example

Reviewer: Do we have task for this?

Author: >mark it TODO for yourself and will add within this PR

this

            }
            Some(_) | None => unreachable!("expected Tokenize response"),
        }
    }

    let request = Request::new(BulkRequestBatch {
        requests: vec![
            BulkRequest {
                request: Some(BulkRequestItem::Detokenize(TokenizeRequest {
                    value: Some(TokenizeValue::EmailValue(tokenized_email)),
                    zone_id: zone_id.iter().cloned().collect(),
                    client_id: client_id.iter().cloned().collect(),
                })),
                request_id: vec![0],
            },
            BulkRequest {
                request: Some(BulkRequestItem::Detokenize(TokenizeRequest {
                    value: Some(TokenizeValue::StrValue(tokenized_string)),
                    zone_id: zone_id.iter().cloned().collect(),
                    client_id: client_id.iter().cloned().collect(),
                })),
                request_id: vec![1],
            },
        ],
    });

    eprintln!("DETOKENIZE REQUEST: {:?}", &request.get_ref());

    let response = client.process_bulk(request).await?;

    eprintln!("DETOKENIZE RESPONSE: {:?}", &response.get_ref());

    Ok(())
}
```

### HTTP

> TODO add example of using curl or whatever

Reviewer: Same. Is there a task for it?

Author: Added later in this PR
